File: Dynamic_Programming/Array_DP/House_Robber.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        if(n==1) return nums[0];
        if(n==2) return max(nums[0], nums[1]);
        int l = max(nums[1], nums[0]), sl = nums[0], tmp;
        for(int i=2; i<n; i++) {
            tmp = max(l, sl+nums[i]);
            sl = l;
            l = tmp;
        }
        return l;
    }
};

File: Dynamic_Programming/Array_DP/House_Robber.py
from typing import List

class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        r = [0]*n
        nr = [0]*n
        r[0] = nums[0]
        for i in range(1, n):
            r[i] = nr[i-1]+nums[i]
            nr[i] = max(r[i-1], nr[i-1])
        print(r)
        print(nr)
        return max(r[n-1], nr[n-1])